### ISSUE TITLE
[Bug] Error 1: Previous Definition Different

### ISSUE THREAD

**Issue by Chigusa0w0** [marler8997/clrbridge#1]
When trying to execute the step `rund example.d` while building the project with command `rund minish.d cleanBuildAll`, tons of error spawned saying _Error 1: Previous Definition Different {Mangled Name Here}_.

After reading the codegen source code. I found that currently _clrbridge_ lacks of a way to handle different overloads with only non-primitive parameters have different type.

For example, for [Marshal.ReadInt64](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.readint64?view=netframework-4.8), both of its two-parameter overloads become `ReadInt64(__d.clr.DotNetObject, int)`. Thus a case of the above mentioned _Error 1_ occurred.

Could there be a way to identify and fix this issue? 

A possible way to solve this is to generate _DotNetObject_IntPtr_ for C# _IntPtr_, _DotNetObject_Object_ for C# _Object_ and **alias this** each of these _DotNetObject\_???_ to _DotNetObject_ so that the line `__d.globalClrBridge.box!"Object"` could leave unchanged.

e.g.:
 `ReadInt64(IntPtr, Int32)` -> `ReadInt64(__d.clr.DotNetObject_IntPtr, int)`
 `ReadInt64(Object, Int32)` -> `ReadInt64(__d.clr.DotNetObject_Object, int)`

Also worth mentioning, as you may know, [not everything in C# derives from object](https://docs.microsoft.com/en-us/archive/blogs/ericlippert/not-everything-derives-from-object). So, although I haven't tested this by myself, try calling some [String Constructors](https://docs.microsoft.com/en-us/dotnet/api/system.string.-ctor?view=netframework-4.8) (and other methods with pointer-type or _TypedReference_ parameters) may cause error or unexpected behaviour, thus need to be excluded from codegen procedure.

PS. may I ask what license is applied to this project?

**Comment by marler8997** [comment]
Hey @LimiQS thanks for trying out the project.

I started this project a month ago and am aware of the issue with overloads.  I should have that part of the project implemented soon.  The reason for the error is just that I'm currently skipping code generation for some type definitions for various reasons, so I'm just treating all type references as Objects.  So the current types are just placeholders until it gets further along.

**Comment by marler8997** [comment]
Update: I've made some more progress on the project and have found a solution for overload problem during this initial development stage.  I've added an `UnuspportedType` template which is implicitly convertible to `DotNetObject`, but uses a template parameter so each instance will have a unique type name to avoid the overload problem until all types are supported.

As to what license this is, that's a great question!  The development of this project is being sponsored by Symmetry Investments, so I will ask them what license they'd like me to use here.  I'm of the opinion that when you have a very "general" library like this, it's better to use open licenses like MIT/BOOST/Apache.  I'll check if they are ok with that.
